### PR TITLE
6665 publicSubscribers

### LINKED ISSUE #6665: Provide way for user to read their vaults via RPC
## What is the Problem Being Solved?

Currently all the vaults are readable via RPC, for example `published.vaultFactory.manager0.vaults.vault0`. But there is no way for a dapp, given a user's smart wallet address, to know which vaults belong to that user.

## Description of the Design

using `publicSubscribers`, put the storage paths into RPC

## Security Considerations

No concerns with publishing this data because it's public information on the chain.

## Test Plan

A working CLI would be a good proof of concept. Something like:
```
$ agops vaults query --wallet agoric1xyz

> [manager0.vault0, manager0.vault4, manager4.vault7]
```
Also, would be good to confirm that the asset subscriber rpc in https://github.com/Agoric/agoric-sdk/blob/master/packages/inter-protocol/src/vaultFactory/vaultKit.js#L21-L25 can be discovered based on the vault id.

### PR DESCRIPTION
**Pull request by turadg**
closes: #6665

## Description

RPC clients need to be able to query what vaults they've opened. For the identity, since it's over vstorage, we have to use the smart wallet. To keep the smart wallet decoupled from any particular contracts, we need a general mechanism for contracts to return data that should be put into vstorage.

The solution here is to use the `publicSubscribers` pattern in the offer result, and store this in a new datum of the `.current` wallet storage node. Each StoredSubscriber gets represented by its `getStoreKey()` which is a JSON that the Casting lib can consume to read the efed.

E.g. `:published.wallet.agoric109q3uc0xt8aavne94rgd6rfeucavrx924e0ztf.current` includes a value like,
```
{
  "offerToPublicSubscriberPaths": {
    "1674172865518": {
      "asset": {
        "dataPrefixBytes": "\x00",
        "storeName": "vstorage",
        "storeSubkey": "3\x00published\x00vaultFactory\x00manager0"
      },
      "vault": {
        "dataPrefixBytes": "\x00",
        "storeName": "vstorage",
        "storeSubkey": "5\x00published\x00vaultFactory\x00manager0\x00vaults\x00vault1"
      }
    }
  }
}
```

To test the system, this also adds `agops vaults` commands,
- `open` prepares an offer to open a vault
- `list` reads RPC to show a list of vaults owned by the specified smart wallet

I pretty carefully curated the commits so you may want to review by commit.

### Security Considerations

Publishes the `publicSubscribers` of an offer result. This is ostensibly what they're for. The offers themselves were already public in chain history.

### Documentation Considerations

Not sure. This is a new part of the client API but maybe the CLI commands are sufficient?

### Testing Considerations

Manually run through `agops-vaults-smoketest.sh`. It's infeasible to automate because of all that it integrates.

This currently has an XXX commit to enable extra logging and start each VaultManager with a high `debtLimit`.

### CODE REVIEW

File: packages/casting/src/casting-spec.js [hunk 1]
```diff
@@ -28,19 +29,40 @@ const vstoragePathToCastingSpec = (storagePath, storeName = 'vstorage') => {
...
-    storeSubkey: buf.map(b => (b === PATH_SEPARATOR_BYTE ? 0 : b)),
+    storeSubkey: buf.map(b =>
+      b === PATH_SEPARATOR_BYTE ? KEY_SEPARATOR_BYTE : b,
+    ),
     dataPrefixBytes: DATA_PREFIX_BYTES,
   });
 };
 
+// TODO make a similar castingSpecToPath.
+// This one uses VStorageKey which has storeSubkey:string, not UInt8Array
```
Comment (turadg) [inline packages/casting/src/casting-spec.js:40]: @michaelfig it turned out that the `storeSubkey` here was "string", not UInt8Array as in a legit CastingSpec. So I wrote `vstorageKeySpecToPath` and left this note with your idea to have `castingSpecToPath`.

File: packages/agoric-cli/src/lib/vaults.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ * @returns {BridgeAction}
+ */
+export const makeOpenSpendAction = (instance, brands, opts) => {
+  const proposal = makeOpenProposal(brands, opts);
+
+  console.warn('vaults open give', proposal.give);
+  console.warn('vaults open want', proposal.want);
+
+  // NB: not really a Proposal because the brands are not remotes
+  // Instead they're copyRecord like  "{"boardId":"board0257","iface":"Alleged: IST brand"}" to pass through the boardId
```
Comment (dckc) [inline packages/agoric-cli/src/lib/vaults.js:46]: I think making them remotes would be a matter of tweaking their prototype or adding a magic property... looks like `Far` or [Remotable](https://github.com/endojs/endo/blob/master/packages/pass-style/src/make-far.js#L78) would do it.

Comment (dckc) [reply packages/agoric-cli/src/lib/vaults.js:46]: looks like changing `harden({ [slotKey]: slot, iface })` to `Far(iface, { [slotKey]: slot, iface })` would do it:

https://github.com/Agoric/agoric-sdk/blob/c2cedde989634151fc8083ee8b45484860dc0f3a/packages/agoric-cli/src/lib/rpc.js#L193

perhaps modulo getting rid of a redundant `Alleged: `.

File: packages/inter-protocol/src/vaultFactory/vaultManager.js [hunk 3]
```diff
@@ -940,6 +940,8 @@ export const vivifyVaultManagerKit = baggage => {
             want: { Minted: null },
           });
 
+          // NB: This increments even when a vault fails to init and is removed
+          // from the manager, creating a sparse series of published vaults.
```
Comment (dckc) [inline packages/inter-protocol/src/vaultFactory/vaultManager.js:944]: ouch... ISTR from a security audit that erroneous inputs are not supposed to cause any change in state.

Well, better to have it documented, I suppose.

File: packages/agoric-cli/src/commands/vaults.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ */
+export const makeVaultsCommand = async logger => {
+  const vaults = new Command('vaults').description('Vault Factory commands');
+
+  vaults
+    .command('open')
+    .description('open a new vault')
+    .option('--giveCollateral [number]', 'Collateral to give', Number, 9.0)
+    .option('--wantMinted [number]', 'Minted wants', Number, 6.0)
+    .option('--offerId [number]', 'Offer id', Number, Date.now())
```
Comment (dckc) [inline packages/agoric-cli/src/commands/vaults.js:65]: ambient access to the clock. sigh.

File: packages/agoric-cli/src/commands/vaults.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+/** @typedef {import('@agoric/smart-wallet/src/offers').OfferSpec} OfferSpec */
+/** @typedef {import('@agoric/smart-wallet/src/offers').OfferStatus} OfferStatus */
+/** @typedef {import('@agoric/smart-wallet/src/smartWallet').BridgeAction} BridgeAction */
+
+const {
+  vstorage,
+  fromBoard: _fromBoard,
+  agoricNames,
+} = await makeRpcUtils({ fetch });
```
Comment (dckc) [inline packages/agoric-cli/src/commands/vaults.js]: ambient access to the network. sigh.

File: packages/smart-wallet/src/smartWallet.js [hunk 6]
```diff
@@ -277,6 +286,9 @@ const SmartWalletKit = defineVirtualFarClassKit(
             balance: a,
           })),
           offerToUsedInvitation: mapToRecord(offerToUsedInvitation),
+          offerToPublicSubscriberPaths: mapToRecord(
+            offerToPublicSubscriberPaths,
+          ),
```
Comment (dckc) [inline packages/smart-wallet/src/smartWallet.js:293]: hm... I mostly see maps shared in `.entries()` form. Too bad I didn't think of that for `offerToUsedInvitation`.

File: packages/agoric-cli/src/commands/vaults.js [hunk 7]
```diff
@@ -2,28 +2,61 @@
...
  */
 export const makeVaultsCommand = async logger => {
-  const vaults = new Command('vaults').description('Vault Factory commands');
+  const vaults = new Command('vaults')
+    .description('Vault Factory commands')
+    .option('--home [dir]', 'agd application home directory')
+    .option(
+      '--keyring-backend [os|file|test]',
+      'keyring\'s backend (os|file|test) (default "os")',
+    );
```
Comment (dckc) [inline packages/agoric-cli/src/commands/vaults.js:29]: why keyring options? we're not signing, are we?

ah... we're looking up addresses by name. got it.

File: packages/agoric-cli/src/commands/vaults.js [hunk 8]
```diff
@@ -2,28 +2,61 @@
...
+      'list vaults ever owned by the address (as path that can be followed)',
+    )
+    .requiredOption(
+      '--from <address>',
+      'wallet address literal or name',
+      normalizeAddress,
+    )
+    .action(async function () {
+      // @ts-expect-error this implicit any
+      const opts = this.opts();
```
Comment (dckc) [inline packages/agoric-cli/src/commands/vaults.js:46]: is `opts` available as an argument to the action function?

File: packages/inter-protocol/src/proposals/addAssetToVault.js [hunk 9]
```diff
@@ -262,7 +262,11 @@ export const addAssetToVault = async (
   const stable = await stableP;
   const vaultFactoryCreator = E.get(vaultFactoryKit).creatorFacet;
   await E(vaultFactoryCreator).addVaultType(interchainIssuer, oracleBrand, {
-    debtLimit: AmountMath.make(stable, 0n),
+    // DONOTMERGE used for testing, but this script will run in production.
+    // Must the debtLimit start at zero?
```
Comment (dckc) [inline packages/inter-protocol/src/proposals/addAssetToVault.js]: In the PSM release, we decided it was cost-effective to have a >0 default debtLimit, provided it represents a manageable economic risk. We used 1000IST there.

https://github.com/Agoric/agoric-sdk/blob/bdf9cc9c914227990bd6af0030923290d11603d8/packages/inter-protocol/src/proposals/startPSM.js#L59

Even 250IST or 50IST might suffice for some vaults testing.

The `addAssetToVault` function should have a `config` argument that can easily be used by a proposer to the BLDer DAO to choose the initial debt limit.

**Review by dckc — COMMENTED** [review]
looks pretty good.

I'll hold off approval, given the final DO NOT MERGE commit.

I haven't run the integration test scripts; I hope @samsiegart beats me to it.

I'd like to see @michaelfig approve the casting stuff. Otherwise I should look at it when I'm more fresh; some documentation might help (or did I miss it?)

File: packages/inter-protocol/scripts/start-local-chain.sh [hunk 10]
```diff
@@ -58,7 +58,8 @@ make fund-provision-pool
 echo "Funding your wallet account..."
 # After `fund-provision-pool` there is 900 IST remaining for other account funding.
 # A wallet can be tested with 20 BLD for provisioning wallet and 20 USDC for psm trading
-make ACCT_ADDR="$WALLET_BECH32" FUNDS=20000000ubld,20000000ibc/usdc1234 fund-acct
+# Also include 1M IbcATOM
+make ACCT_ADDR="$WALLET_BECH32" FUNDS=20000000ubld,20000000ibc/usdc1234,1000000000000ibc/atom1234 fund-acct
```
Comment (samsiegart) [inline packages/inter-protocol/scripts/start-local-chain.sh:71]: 🥇

File: packages/agoric-cli/test/agops-vaults-smoketest.sh [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+# which is required for the maxDebtFor check that VM makes before opening a vault.
+
+# open a vault
+OFFER=$(mktemp -t agops.XXX)
+bin/agops vaults open >|"$OFFER"
+jq ".body | fromjson" <"$OFFER"
+agoric wallet send --keyring-backend="test" --from "$WALLET" --offer "$OFFER"
+
+# list my vaults
+bin/agops vaults list --from dev-wallet
```
Comment (samsiegart) [inline packages/agoric-cli/test/agops-vaults-smoketest.sh:39]: Confirmed this command works

**Review by samsiegart — COMMENTED** [review]
I got this loading in the UI with https://github.com/Agoric/agoric-sdk/commit/76d3a7d72cfc796d5e8b8edd0e75bb3d03dee904 and https://github.com/Agoric/dapp-inter/commit/c860dbc22ef712407c1204100e9d7afd2c866c02.

Two concerns:
- Can we publish this info in the latest updates in addition to just `.current`? Maybe we could include it in offer status updates? Right now the UIs are set up to read `.current` once then follow latest, so that'd be convenient.
- Is there a convenient way to parse `"storeSubkey": "3\x00published\x00vaultFactory\x00manager0"` that I don't know about? It worked great for passing right into `casting`, but I'd also like to parse it to infer the corresponding vault manager and check to see that the offer is in fact a vault creation offer.

File: packages/agoric-cli/src/commands/vaults.js (hunk 5 above)
Comment (turadg) [reply packages/agoric-cli/src/commands/vaults.js]: `makeRpcUtils` takes `fetch` to make that an ocap.

Think of each command file as a script start. It's gathering the authorities. This once comes from global.

File: packages/inter-protocol/src/proposals/addAssetToVault.js (hunk 9 above)
Comment (turadg) [reply packages/inter-protocol/src/proposals/addAssetToVault.js]: SGTM, I'll make it 1000IST like startPSM. 

I agree it should be a config option. I wrestled a bit with introducing it so I propose it be created when needed.

**Comment by turadg** [comment]
> Can we publish this info in the latest updates in addition to just .current? Maybe we could include it in offer status updates? Right now the UIs are set up to read .current once then follow latest, so that'd be convenient.

Hmm. We could add an `UpdateRecord` publish for that. We didn't for invitationMakers though. Chain storage is expensive and if the client can get what it needs without additional writes we should do that. I think the UIs should continue to read `.current` and if that's too much coding we can make a helper.

The reason we have the UpdateRecord at all is to get intermediate states like `offerStatus` changes without publishing a new `.current`. But when we complete a continuing offer we do publish a new `.current` so any other output would be redundant.

> Is there a convenient way to parse "storeSubkey": "3\x00published\x00vaultFactory\x00manager0" that I don't know about? It worked great for passing right into casting, but I'd also like to parse it to infer the corresponding vault manager and check to see that the offer is in fact a vault creation offer.

It's new in this PR: https://github.com/Agoric/agoric-sdk/blob/0cffca854b405fac48423aba06c82a51c3dc5883/packages/casting/src/casting-spec.js#L46-L56
Though I need to confer with @michaelfig on whether we could just store that string in chainstorage.

**Review by samsiegart — APPROVED** [review]
LGTM after having tested e2e with dapps

File: packages/agoric-cli/src/commands/vaults.js (hunk 8 above)
Comment (turadg) [reply packages/agoric-cli/src/commands/vaults.js:46]: looks like somewhere along the way they added that https://www.npmjs.com/package/commander#action-handler

I'll tackle that in upgrading Commander to v10

**Comment by turadg** [comment]
After rebasing on master I did the manual test again and it's failing now to open a vault,
```
2023-01-20T23:04:49.108Z SwingSet: vat: v27: wallet agoric109q3uc0xt8aavne94rgd6rfeucavrx924e0ztf starting executeOffer 1674255888210
2023-01-20T23:04:49.275Z SwingSet: ls: v19: Logging sent error stack (Error#2)
2023-01-20T23:04:49.276Z SwingSet: ls: v19: Error#2: Withdrawal of { brand: Object [Alleged: IbcATOM brand] {}, value: 9_000_000n } failed because the purse only contained { brand: Object [Alleged: IbcATOM brand] {}, value: 0n }
```

However the account does have the funds,
```
$ agd query bank balances "$WALLET_BECH32"
balances:
- amount: "1000000000000"
  denom: ibc/atom1234
- amount: "20000000"
  denom: ibc/usdc1234
- amount: "10000000"
  denom: ubld
- amount: "250000"
  denom: uist
```

So I think it's failing due to https://github.com/Agoric/agoric-sdk/pull/6774 . Assuming this passes CI I'll land this and work separately to resolve the new integration problem.

**Comment by mhofman** [comment]
@mergifyio requeue

**Pull request merged**